src/sqlfluff/testing/rules.py

```diff
@@ -120,6 +120,42 @@ def assert_rule_raises_violations_in_file(rule, fpath, violations, fluff_config)
     assert set(lnt.check_tuples()) == {(rule, v[0], v[1]) for v in violations}
 
 
+def prep_violations(rule, violations):
+    """Default to test rule if code is omitted."""
+    for v in violations:
+        if "code" not in v:
+            v["code"] = rule
+    return violations
+
+
+def assert_violations_before_fix(test_case, violations_before_fix):
+    """Assert that the given violations are found in the given sql."""
+    violation_info = [e.get_info_dict() for e in violations_before_fix]
+    try:
+        assert violation_info == prep_violations(test_case.rule, test_case.violations)
+    except AssertionError:
+        print("Actual violations:\n" + yaml.dump(violation_info))
+        raise
```

Comment:
This is a hack though.

I wanted to simply write this:
```suggestion
    assert violation_info == prep_violations(test_case.rule, test_case.violations), \
        "Actual violations:\n" + yaml.dump(violation_info))
```

But it seems that the assertion message is hidden by the outer layers of the test framework 🤔 

I should figure out how to include the assertion message in output when the assertion fails.

But I suggest leaving that for later. This code works and can be merged, in order to migrate the problematic existing test cases ie. enforce explicit assertions for any test cases where violations remain after the fix.

Reply:
OK but then can we add a `  # pragma: no cover` to these lines so test coverage stays at 100%? I'm a bit nerdy about that, cause when less than 100% it's more difficult to see if new code has missing coverage, or rounding errors are causing a slight change. With 100% (even if some are `pragma: no cover` so not really 100%) makes it cleaner and easier to review code.

Reply:
> I should figure out how to include the assertion message in output when the assertion fails.

I tried to find out about this but I couldn't. pytest seems to include only the 1st line of the assert message.

Printing the actual violations in plain yaml format seems so useful, that I added the "hack" back.

Reply:
Do you see the full output if you run with `-vv`? I have often used that to see additional output (e.g. full text diff when expected and actual SQL do not match).

https://docs.pytest.org/en/latest/how-to/output.html

Reply:
@barrywhart `-vv` helps indeed – my bad 🤦‍♂️

Btw I'm running a single test with this command to test it:
```
pytest -s test/rules/yaml_test_cases_test.py -k L004_indented_comments_tab_config -vv
```

If my assert code is:
```
    assert violation_info == prep_violations(
        test_case.rule, test_case.violations_after_fix
    ), "Actual violations:\n" + yaml.dump(violation_info)
```

I can find this in the output:

> E       AssertionError: Actual violations:
E         - code: L004
E           description: Incorrect indentation type found in file. The indent occurs after other
E             text, so a manual fix is needed.
E           line_no: 2
E           line_pos: 4
E
E       assert [{'code': 'L004',\n  'description': 'Incorrect indentation type found in file. The indent occurs '\n                 'after other text, so a manual fix is needed.',\n  'line_no': 2,\n  'line_pos': 4}] == [{'code': 'L004',\n  'description': 'Incorrect indentation type found in file. The indent occurs '\n                 'after other text, so a manual fix is needed.',\n  'line_no': 2,\n  'line_pos': 2}]
E         At index 0 diff: {'line_no': 2, 'line_pos': 4, 'code': 'L004', 'description': 'Incorrect indentation type found in file. The indent occurs after other text, so a manual fix is needed.'} != {'code': 'L004', 'description': 'Incorrect indentation type found in file. The indent occurs after other text, so a manual fix is needed.', 'line_no': 2, 'line_pos': 2}
E         Full diff:
E           [
E            {'code': 'L004',
E             'description': 'Incorrect indentation type found in file. The indent occurs '
E                            'after other text, so a manual fix is needed.',
E             'line_no': 2,
E         -   'line_pos': 2},
E         ?               ^
E         +   'line_pos': 4},
E         ?               ^
E           ]

Having those lines prefixed with "E" is not friendly compared to the custom print (current test code) that is:

```
Actual violations_after_fix:
- code: L004
  description: Incorrect indentation type found in file. The indent occurs after other
    text, so a manual fix is needed.
  line_no: 2
  line_pos: 4
```

..which can be copied as is into yaml file. But that is even harder to find in the test output because it's printed some lines before the AssertionError :/ So for console output there is no perfect option.

Also, in the console the AssertionError printed quite early and my screen gets filled with lots of debug logging after it. Basically the last thing that is logged is this:

> FAILED test/rules/yaml_test_cases_test.py::test__rule_test_case[L004_indented_comments_tab_config] - AssertionError: assert [{'code': 'L004',\n  'description': 'Incorrect indentation type found in file. The indent occurs '\n                 'after o...

So it's hard to find the actual diff from the console output (Also need to realize to add `-vv`). I tried the same code with my IDE and it's basically nicer because it basically just shows the diff. But with that code my IDE entirely hides the "Actual violations_after_fix" message! It only shows me the string comparison.

The custom print (current code) works very nicely in IDE – there I basically only the important output:

```
Actual violations_after_fix:
- code: L004
  description: Incorrect indentation type found in file. The indent occurs after other
    text, so a manual fix is needed.
  line_no: 2
  line_pos: 4

FAILED
test/rules/yaml_test_cases_test.py:17 (test__rule_test_case[L004_indented_comments_tab_config])
[{'code': 'L004',
  'description': 'Incorrect indentation type found in file. The indent occurs '
                 'after other text, so a manual fix is needed.',
  'line_no': 2,
  'line_pos': 4}] != [{'code': 'L004',
  'description': 'Incorrect indentation type found in file. The indent occurs '
                 'after other text, so a manual fix is needed.',
  'line_no': 2,
  'line_pos': 2}]

<Click to see difference>
```

To conclude, at least for my workflow the custom print option works best.